src/attribute.ts

```diff
@@ -66,60 +73,116 @@ export class Attribute {
+  static from(serialized: any): Attribute {
+    const attr = new Attribute(serialized.definition)
+    attr.isItem = serialized.isItem
+    attr.index = serialized.index
+    for (const child of serialized.children) {
+      attr.addChild(Attribute.from(child))
+    }
+    for (const slice of serialized.slices) {
+      attr.addSlice(Attribute.from(slice))
+    }
+    for (const item of serialized.items) {
+      attr.addItem(Attribute.from(item))
+    }
+    return attr
+  }
 
-    if (this.definition.sliceName && this.parent?.name.includes('[x]')) {
-      return this.definition.sliceName
+  spreadTypes(): Attribute[] {
+    if (this.types.length > 1) {
+      return this.types.map(type => {
+        const attr = new Attribute({
+          ...this.definition,
+          type: [{ code: type }],
+          id: this.definition.id.replace('[x]', toCamelCase(type)),
+          path: this.definition.path.replace('[x]', toCamelCase(type)),
+        })
+        this.parent?.addChild(attr)
+        return attr
+      })
     }
+    return [this]
+  }
 
-    // if the parent has multiple types, use the type in camelCase
-    if (this.parent && this.parent.types.length > 1)
-      return this.name.replace('[x]', toCamelCase(this.definition.type[0].code))
+  isChild(p: Attribute): boolean {
+    let current = this as Attribute | undefined
+    while (current) {
+      if (current.equals(p)) return true
+      current = current.parent
+    }
+    return false
+  }
 
-    return this.name || ''
+  addChild(child: Attribute) {
+    child.parent = this
+    this.children.push(child)
   }
 
-  serialize(): string {
-    // if not parent, return the definitionId
-    if (!this.parent) {
-      return this.tail()
+  addSlice(slice: Attribute) {
+    slice.parent = this.parent
+    if (this.isItem) {
+      slice.isItem = true
+      slice.index = this.index
     }
+    this.slices.push(slice)
+  }
 
-    if (this.parent.isArray) {
-      return `${this.parent.serialize()}${this.tail()}`
+  addItem(item: Attribute, index?: number) {
+    const computeIndex = () => {
+      if (index !== undefined) {
+        if (this.items.map(it => it.index).includes(index)) {
+          throw new Error(`item with index ${index} already exists`)
+        }
+        return index
+      }
+
+      for (let i = 0; i <= this.items.length; i++) {
+        if (!this.items.map(it => it.index).includes(i)) return i
+      }
+      return this.items.length
     }
 
-    // if parent is a multi-type attribute, we don't want to render the parent
-    if (
-      this.parent.types.length > 1 ||
-      (this.parent.definition.slicing && this.parent.name.includes('[x]'))
-    ) {
-      return this.parent.parent
-        ? `${this.parent.parent.serialize()}.${this.tail()}`
-        : this.tail()
-    }
+    item.parent = this.parent
+    item.isItem = true
+    item.index = index !== undefined ? index : computeIndex()
+    item.slices.forEach(slice => {
```

Comment:
yeah I didn't explain shit that's why, i'll add comments.
Basically what I do here is that when an item (with slices) is added to an array, we must tell the slices that they are in fact items of the array themselves. If you have for instance `Observation.category` which is an array, and you want to add a "sliced children" like `Observation.category:VSCat` to the array, you need to put the index in the slice so that the path is computed in `category[index]` rather than just `category`